### ISSUE TITLE
Comments: Mutexes in Rust and C++: Protecting Data versus Protecting Code

### ISSUE THREAD

**Issue by geo-ant** [geo-ant/blog#17]
Comments for this article go here.

**Comment by ChenZhongPu** [comment]
In Rust, The `mutex` is used to protect some data. So what is the idiomatic way to protect a struct?

Here is the code from [cmu-bustub](https://github.com/cmu-db/bustub), where class [DiskManager](https://github.com/cmu-db/bustub/blob/master/src/include/storage/disk/disk_manager.h) is shipped with a `Mutex`. How can we implement this in Rust? 

In my opinion, C++'s mutex here is to protect both `db_io_` and `log_io_`, so we should create two Mutexes for them, respectively. Am I right?

**Comment by geo-ant** [comment]
Hi @ChenZhongPu, I skimmed the code and it seems the `db_io_latch_` mutext is locked whenever the `db_io_` member is manipulated. This is actually a brilliant example of what I think the difference between the Rust and C++ logic for mutexes is. In the C++ code you have to remember to lock the `db_io_latch_` mutex every time you want to interact with `db_io_`. There is nothing the compiler can do to help you to do that.

Let's look how equivalent Rust code could look. A rust mutex has ownership over the data that it protects and you can only access it when you unlock the mutex. So in the C++ code you want to protect an instance of `fstream`, which is roughly equivalent to Rusts`s `File. So in idiomatic Rust, you would not have two member fields `db_io_`and the associated mutex `db_io_latch`, you would just have one member of type `std::sync::Mutex<std::fs::File>`. Now you can only access the content (i.e. the file stream) of that mutex when you call `lock`, which gives you a thing that is similar to a scoped lock that can be dereferenced into the file object you are interested in. So in this case the data that you want to protect is the filestream itself. There is no way you can forget to lock the mutex because otherwise you cannot access the data behind that mutex.

**Note for completeness**: if you want to replicate the C++ code, then you could always create a `std::sync::Mutex<()>`, which carries no data to protect. This would have the same semantics as the C++ mutex and you would have to remember to lock it every time you want to interact with your filestream. But this loses all the good stuff that idiomatic usage would give you.

**Comment by ChenZhongPu** [comment]
I am a bit confused. As for `Mutex<()>`, if it is not to protect any data, then will it protect the "code" (like what C++ did)？ 

BTW, I also noticed that the C++ code didn't introduce any mutex (or lock) for log I/O? Is there any thread safety problem here?

**Comment by geo-ant** [comment]
> I am a bit confused. As for Mutex<()>, if it is not to protect any data, then will it protect the "code" (like what C++ did)？

Exactly. There are use cases for that, but the example you gave illustrates nicely why the idiomatic use of Rust mutexes is great.

> BTW, I also noticed that the C++ code didn't introduce any mutex (or lock) for log I/O? Is there any thread safety problem here?

Without having looked at the code, the answer is likely yes. Unless your logging backend has some sort of synchronization, you'll might end up with weirdly formatted locks. If, by chance, the `db_io_latch_` mutex is always locked when logging, there would of course be no problem, but there might be one in the future :)

**Comment by jean-marc** [comment]
Thanks for this article, I think the formula 'protecting data vs code' is on point, the nice thing about C++ is that one can 'piggy back' on a mutex: since it's already locked we might as well do multiple operations, that's not as simple in Rust.
There seems to be some overhead in Rust because two synchronizations are taking place: for the counter _and_ the reference counter. I understand that the reference counter uses atomic variable which is lighter than a mutex. Have you compared performance between the Rust and C++ version? Cheers

**Comment by geo-ant** [comment]
Hey thanks, no I've never compared the performance of a Rust `Mutex<T>` and something that's protected via a C++ `std::mutex`. I'm sure there are useful ways of doing that, I just never did. Having said that, I'd be surprised if there's a meaningful difference. 

Just note that `Arc<Mutex<T>>` isn't the only way of using a mutex in Rust. It's just a simple way of making data both shareable and mutex protected. In C++ you'd also have to share data between threads somehow. Of course you could just use references because CPP doesn't enforce thread safety at all. But if you want lifetime management then you need `std::shared_ptr`. At this point, you have the CPP equivalent of `Arc<Mutex<T>>`. Shared references are a bit tougher to do in Rust if you just use regular threads (because of `'static` bounds), but scoped threads work well with references. All I'm saying is that `Arc<Mutex<T>>` isn't the only way of sharing mutex protected data. 

I'm not quite sure what you mean with the piggy-backing on a locked mutex. In Rust you can definitely use the `lock` member function. If you assign the result to a named variable, this gives you an RAII [`MutexGuard`](https://doc.rust-lang.org/std/sync/struct.MutexGuard.html) which lets you perform many operations on the protected data. Once this guard goes out of scope, the mutex is locked again. For example:

```rust
{
  let mut handle = 
    my_mutex
    .lock()
    .unwrap();

  handle.add(20);
  handle.multiply(10);
  // once handle goes out
  // of scope, the mutex
  // is locked again...
}
```

The one thing that's genuinely harder to do in Rust is locking multiple mutexes without deadlocking, like `std::lock` or `std::lock_guard` do in CPP. I don't know a standard library way of doing that (it might exist). But I think there are crates for that.
